app/controllers/users.js

```diff
@@ -1,23 +1,42 @@
-const { createUser, findAllUsers, findUserByEmail } = require('../services/users');
+const { badRequestError } = require('../errors');
+const { RolesType } = require('../fixtures/roles');
 const logger = require('../logger');
-const { mapToSerializer } = require('../utils/objects');
 const { signUpSerializer, signInSerializer, getUsersSerializer } = require('../serializers/users');
+const { createUser, findAllUsers, findUserByEmail } = require('../services/users');
 const { isOriginalText } = require('../utils/crypto');
 const { signPayload } = require('../utils/jwt');
+const { mapToSerializer } = require('../utils/objects');
 const { accessTokenExpirationTime } = require('../../config').common.session.times;
-const { badRequestError } = require('../errors');
 
 const loggerPath = 'controller:users';
 
 exports.signUp = async (req, res, next) => {
   try {
+    logger.info(`${loggerPath}:signUp: role is ${res.locals.user.role}`);
     const { body: userData } = req;
+    userData.role = res.locals.user.role;
     logger.info(
       `${loggerPath}:signUp: starting signUp method with the next body ${JSON.stringify(userData)}`
     );
 
-    const createdUser = await createUser(userData);
-    return res.status(201).send(mapToSerializer(createdUser, signUpSerializer));
+    const foundUser = await findUserByEmail(userData.email);
+    logger.info(`${loggerPath}:signUp: found user by email is ${foundUser}`);
+
+    const wantToBeAdmin = userData.role === RolesType.ADMIN;
+
+    let user = null;
+
+    if (foundUser) {
+      const isAdmin = foundUser.role === RolesType.ADMIN;
+      if ((isAdmin && wantToBeAdmin) || !wantToBeAdmin) throw badRequestError('email already exists');
+
+      foundUser.role = RolesType.ADMIN;
+      user = await foundUser.save();
+    } else {
+      user = await createUser(userData);
+    }
+
+    return res.status(201).send(mapToSerializer(user, signUpSerializer));
```

Comment:
It seems to me that it is much cleaner to separate these processes, a controller to continue creating regular users and another to create or update a user if it exists, you can do this using the upsert function of the user model.